firmware/disk_drive.c

```diff
@@ -466,6 +473,196 @@ static void disk_handle_load(D64 *d64, char *filename, uint8_t *ptr)
+        if((uint16_t)recv_size + save_buffer->data_sector.pos > D64_SECTOR_DATA_LEN) {
+            uint8_t remaining = D64_SECTOR_DATA_LEN - save_buffer->data_sector.pos;
+            if(remaining>0)
+            {
+                memcpy(&save_buffer->data_sector.d64sector.data[save_buffer->data_sector.pos],
+                       receive_buffer,
+                       remaining);
+            }
+            if(!d64_get_next_free_sector(save_buffer, 
+                                         save_buffer->data_sector.track, save_buffer->data_sector.sector, 
+                                         &next_track, &next_sector)) 
+            {
+                return false;
+            }
+            save_buffer->data_sector.d64sector.next_track = next_track;
+            save_buffer->data_sector.d64sector.next_sector = next_sector;
+            if(!d64_write_sector(d64, &save_buffer->data_sector.d64sector, 
+                                 save_buffer->data_sector.track, save_buffer->data_sector.sector))
+            {
+                return false;
+            }
+            
+            memcpy(&save_buffer->data_sector.d64sector.data[0],
+                    receive_buffer + remaining,
+                    recv_size - remaining);
+            save_buffer->data_sector.pos = recv_size - remaining;
+            save_buffer->data_sector.track = next_track;
+            save_buffer->data_sector.sector = next_sector;
+        }
+        else
+        {
+            memcpy(&save_buffer->data_sector.d64sector.data[save_buffer->data_sector.pos], 
+                   receive_buffer,
+                   recv_size);
+            save_buffer->data_sector.pos += recv_size;
+        }
+        c64_interface(true);
+        c64_send_data("OK", 2);
+        c64_send_reply(recv_size);
+    }
+    save_buffer->data_sector.d64sector.next_track = 0;
+    save_buffer->data_sector.d64sector.next_sector = save_buffer->data_sector.pos+1;
+    memset(&save_buffer->data_sector.d64sector.data[save_buffer->data_sector.pos], 
+            0x00, 
+            D64_SECTOR_DATA_LEN - save_buffer->data_sector.pos);
+
+    c64_interface(false);
+    if(!d64_write_sector(d64, &save_buffer->data_sector.d64sector, 
+                            save_buffer->data_sector.track, save_buffer->data_sector.sector))
+    {
+        return false;
+    }
+    return true;
+}
+
+static void disk_handle_save_prg(D64 *d64, D64_SAVE_BUFFER *save_buffer, char *filename, uint16_t filesize_in_bytes) {
+    PARSED_FILENAME parsed_filename;
+    save_buffer->modified_directory_sector_1.track = 0;
+    save_buffer->modified_directory_sector_2.track = 0;
+    
+    // TODO: implement D71 and D81 image type support
+    if(d64->image_type!=D64_IMAGE_D64) 
+    {
+        c64_send_reply(REPLY_NOT_SUPPORTED);
+        return;
+    }
+   
+    disk_parse_filename(filename, &parsed_filename);
+    if (parsed_filename.drive!=0)               // this is a 'single drive system'
```

Comment:
I'm a bit confused here. I thought in the 'd:filename' notation the 'd' meant the drive number, in case of a dual drive system. Hence 'd' can be either 0 or 1 - KFF and 1541 defaults to '0'.
In case 1 is provided KFF drops the request and forwards it to the IEC bus. I think either SD2IEC nor a real 1541 could handle that request.

Reply:
SD2IEC, Pi1541 and apparently CMD drives uses drive numbers for accessing multiple partitions. The default partition is 1, so even if you only have one partition you can still access the content of the SD card using drive number 0 or 1. The SD2IEC does also support a special partition for hardware configurations.
So a SD2IEC can handle a request to drive number 1, or higher depending on the number of partitions. A real 1541 would only respond to drive number 0.